cpp/src/arrow/json/reader.cc

```diff
@@ -42,132 +42,476 @@ namespace arrow {
+ public:
+  explicit DecodingOperator(std::shared_ptr<const DecodeContext> context)
+      : context_(std::move(context)) {}
 
-        // Get all whole objects entirely inside the current buffer
-        RETURN_NOT_OK(chunker_->Process(starts_with_whole, &whole, &next_partial));
-      }
+  Result<DecodedBlock> operator()(const ChunkedBlock& block) const {
+    int64_t num_bytes;
+    ARROW_ASSIGN_OR_RAISE(auto unconverted, ParseBlock(block, context_->parse_options(),
+                                                       context_->pool(), &num_bytes));
 
-      // Launch parse task
-      task_group_->Append([self, partial, completion, whole, block_index] {
-        return self->ParseAndInsert(partial, completion, whole, block_index);
-      });
-      block_index++;
+    std::shared_ptr<ChunkedArrayBuilder> builder;
+    RETURN_NOT_OK(MakeChunkedArrayBuilder(TaskGroup::MakeSerial(), context_->pool(),
+                                          context_->promotion_graph(),
+                                          context_->conversion_type(), &builder));
+    builder->Insert(0, field("", unconverted->type()), unconverted);
 
-      partial = next_partial;
-      block = next_block;
-    }
+    std::shared_ptr<ChunkedArray> chunked;
+    RETURN_NOT_OK(builder->Finish(&chunked));
+    ARROW_ASSIGN_OR_RAISE(auto batch, RecordBatch::FromStructArray(chunked->chunk(0)));
 
-    std::shared_ptr<ChunkedArray> array;
-    RETURN_NOT_OK(builder_->Finish(&array));
-    return Table::FromChunkedStructArray(array);
+    return DecodedBlock{std::move(batch), num_bytes};
   }
 
  private:
-  Status MakeBuilder() {
-    auto type = parse_options_.explicit_schema
-                    ? struct_(parse_options_.explicit_schema->fields())
-                    : struct_({});
+  std::shared_ptr<const DecodeContext> context_;
+};
 
-    auto promotion_graph =
-        parse_options_.unexpected_field_behavior == UnexpectedFieldBehavior::InferType
-            ? GetPromotionGraph()
-            : nullptr;
+// TODO(benibus): Replace with `MakeApplyGenerator` from
+// github.com/apache/arrow/pull/14269 if/when it gets merged
+//
+// Reads from the source and spawns fan-out decoding tasks on the given executor
+AsyncGenerator<DecodedBlock> MakeDecodingGenerator(
+    AsyncGenerator<ChunkedBlock> source,
+    std::function<Result<DecodedBlock>(const ChunkedBlock&)> decoder,
+    Executor* executor) {
+  struct State {
+    AsyncGenerator<ChunkedBlock> source;
+    std::function<Result<DecodedBlock>(const ChunkedBlock&)> decoder;
+    Executor* executor;
+  } state{std::move(source), std::move(decoder), executor};
+
+  return [state = std::make_shared<State>(std::move(state))] {
+    auto options = CallbackOptions::Defaults();
+    options.executor = state->executor;
+    // Since the decode step is heavy we want to schedule it as
+    // a separate task so as to maximize task distribution accross CPU cores
+    options.should_schedule = ShouldSchedule::Always;
```

Comment:
@pitrou After splitting the async-reentrant tests into a separate fixture, the stress tests now get run first. As a result, I'm frequently seeing `DCHECK` output after all the tests have passed - `Invalid: operation forbidden during or after shutdown` at  [future.cc:107](https://github.com/benibus/arrow/blob/25a57e170c5966ddb84d6a4d7de5304c0893cd0a/cpp/src/arrow/util/future.cc#L107). From what I can tell, this is from the threaded tests that exit early with readahead futures still on the global thread pool.

The easiest way to prevent this is to change this line to `ShouldSchedule::Never` (the default) - or just use `MakeMappedGenerator`, i guess - since the source futures have already been transferred to the cpu executor by then. Async-reentrancy would still work, although parallelism would be less aggressive.

Anyways, I can push the changes - but I'm not sure if you had any deeper insights here.

Reply:
Doesn't appear to make a difference, unfortunately

Reply:
Even when doing so on both the IO and CPU thread pools?

Reply:
Yeah, same for the both of them. What _does_ help is making a new thread pool in `SetUp` and using that as the CPU executor - in which case you don't have to explicitly tear anything down.

Reply:
Perhaps you should try to find out which test(s) exactly trigger this issue?

Reply:
Yeah, looks like `IgnoreLeadingEmptyBlocks` and `InferredSchema` trigger it pretty reliably. Looking into it now.